### ISSUE TITLE
Data race detected

### ISSUE THREAD

**Issue by typeless** [go-gitea/gitea#1453]
- Gitea version (or commit ref):
Gitea 1.1.0

- Git version:
git version 2.7.4

- Operating system:
Ubuntu 16.04.2 LTS

- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:
```
Apr  6 09:41:19 DIR-101 gitea[14276]: ==================
Apr  6 09:41:19 DIR-101 gitea[14276]: WARNING: DATA RACE
Apr  6 09:41:19 DIR-101 gitea[14276]: Write at 0x00c421eea368 by goroutine 136:
Apr  6 09:41:19 DIR-101 gitea[14276]:   os/exec.(*Cmd).Wait()
Apr  6 09:41:19 DIR-101 gitea[14276]:       /home/mura/tools/go/src/os/exec/exec.go:439 +0x15d
Apr  6 09:41:19 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeoutPipeline.func1()
Apr  6 09:41:19 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:71 +0x38
Apr  6 09:41:19 DIR-101 gitea[14276]: Previous read at 0x00c421eea368 by goroutine 148:
Apr  6 09:41:19 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeoutPipeline()
Apr  6 09:41:19 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:77 +0x3fe
Apr  6 09:41:19 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeout()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:96 +0xd1
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/code.gitea.io/git.Fsck()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/git.go:78 +0x103
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/models.GitFsck.func1()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/models/repo.go:1943 +0xe4
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/github.com/go-xorm/xorm.(*Session).Iterate()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/go-xorm/xorm/session_iterate.go:35 +0x1a1
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/models.GitFsck()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/models/repo.go:1951 +0x1fe
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.FuncJob.Run()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:92 +0x34
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.(*Cron).runWithRecovery()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:152 +0x68
Apr  6 09:41:20 DIR-101 gitea[14276]: Goroutine 136 (running) created at:
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeoutPipeline()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:72 +0x25c
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeout()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:96 +0xd1
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/code.gitea.io/git.Fsck()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/git.go:78 +0x103
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/models.GitFsck.func1()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/models/repo.go:1943 +0xe4
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/github.com/go-xorm/xorm.(*Session).Iterate()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/go-xorm/xorm/session_iterate.go:35 +0x1a1
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/models.GitFsck()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/models/repo.go:1951 +0x1fe
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.FuncJob.Run()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:92 +0x34
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.(*Cron).runWithRecovery()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:152 +0x68
Apr  6 09:41:20 DIR-101 gitea[14276]: Goroutine 148 (running) created at:
Apr  6 09:41:20 DIR-101 gitea[14276]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.(*Cron).run()
Apr  6 09:41:20 DIR-101 gitea[14276]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:185 +0x661
Apr  6 09:41:20 DIR-101 gitea[14276]: ==================
```

**Comment by typeless** [comment]
At first glance,  it appears the `cmd.ProcessState` is read and simultaneously written within `cmd.Wait` in the other goroutine.

See https://github.com/go-gitea/gitea/blob/master/vendor/code.gitea.io/git/command.go#L71_L77

**Comment by typeless** [comment]
Another data point.
```
Apr 07 14:24:40 DIR-101 gitea[31691]: ==================
Apr 07 14:24:40 DIR-101 gitea[31691]: WARNING: DATA RACE
Apr 07 14:24:40 DIR-101 gitea[31691]: Write at 0x00c421d72ba8 by goroutine 228:
Apr 07 14:24:40 DIR-101 gitea[31691]:   os/exec.(*Cmd).Wait()
Apr 07 14:24:40 DIR-101 gitea[31691]:       /home/mura/tools/go/src/os/exec/exec.go:439 +0x15d
Apr 07 14:24:40 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeoutPipeline.func1()
Apr 07 14:24:40 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:71 +0x38
Apr 07 14:24:40 DIR-101 gitea[31691]: Previous read at 0x00c421d72ba8 by goroutine 135:
Apr 07 14:24:40 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeoutPipeline()
Apr 07 14:24:40 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:77 +0x3fe
Apr 07 14:24:40 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeout()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:96 +0xd1
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/code.gitea.io/git.Fsck()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/git.go:78 +0x103
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/models.GitFsck.func1()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/models/repo.go:1943 +0xe4
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/github.com/go-xorm/xorm.(*Session).Iterate()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/go-xorm/xorm/session_iterate.go:35 +0x1a1
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/models.GitFsck()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/models/repo.go:1951 +0x1fe
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.FuncJob.Run()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:92 +0x34
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.(*Cron).runWithRecovery()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:152 +0x68
Apr 07 14:24:41 DIR-101 gitea[31691]: Goroutine 228 (running) created at:
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeoutPipeline()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:72 +0x25c
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/code.gitea.io/git.(*Command).RunInDirTimeout()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/command.go:96 +0xd1
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/code.gitea.io/git.Fsck()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/git.go:78 +0x103
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/models.GitFsck.func1()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/models/repo.go:1943 +0xe4
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/github.com/go-xorm/xorm.(*Session).Iterate()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/go-xorm/xorm/session_iterate.go:35 +0x1a1
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/models.GitFsck()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/models/repo.go:1951 +0x1fe
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.FuncJob.Run()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:92 +0x34
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.(*Cron).runWithRecovery()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:152 +0x68
Apr 07 14:24:41 DIR-101 gitea[31691]: Goroutine 135 (running) created at:
Apr 07 14:24:41 DIR-101 gitea[31691]:   code.gitea.io/gitea/vendor/github.com/gogits/cron.(*Cron).run()
Apr 07 14:24:41 DIR-101 gitea[31691]:       /home/mura/devel/go/src/code.gitea.io/gitea/vendor/github.com/gogits/cron/cron.go:185 +0x661
Apr 07 14:24:41 DIR-101 gitea[31691]: ==================
```
Followed by the message below every time when doing `git push -f`:
```
Gitea: Internal error
UpdatePublicKey: database is locked
error: failed to push some refs to 'git@example.com/foo/bar.git'
```
P.S. Not sure whether the above two logs are related.

**Comment by lunny** [comment]
resolved by #1470
